File: 브루투포스/계란으로계란치기_16987.java
package 브루투포스;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 계란으로계란치기_16987 {
    static int N, eggCount;
    static Egg[] eggs;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        eggs = new Egg[N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int hp = Integer.parseInt(st.nextToken());
            int atk = Integer.parseInt(st.nextToken());
            eggs[i] = new Egg(hp, atk);
        }

        attackEgg(0);
        System.out.println(eggCount);
    }

    private static void attackEgg(int cur) {
        if (cur == N) {
            int tmp = 0;
            for (int i = 0; i < N; i++) {
                if (eggs[i].hp <= 0) tmp++;
            }
            eggCount = Math.max(eggCount, tmp);
            return;
        }
        if (eggs[cur].hp <= 0) {
            attackEgg(cur + 1);
            return;
        }

        for (int next = 0; next < N; next++) {
            if (cur == next) continue;
            if (eggs[next].hp <= 0) continue;
            int myHp = eggs[cur].hp;
            int eHp = eggs[next].hp;
            eggs[cur].hp -= eggs[next].atk;
            eggs[next].hp -= eggs[cur].atk;
            attackEgg(cur + 1);
            eggs[next].hp = eHp;
            eggs[cur].hp = myHp;
        }
        if (cur == N - 1) attackEgg(cur + 1);
    }

    private static class Egg {
        int hp;
        int atk;

        public Egg(int hp, int atk) {
            this.hp = hp;
            this.atk = atk;
        }
    }
}

File: 코틀린_코테스터디/week19/계란으로계란치기16987.kt
package 코틀린_코테스터디.week19

import java.util.StringTokenizer
import kotlin.math.max

/*
3
8 5
1 100
3 5
 */
class Solution16987 {
    companion object {
        private var N: Int = 0
        private var max: Int = 0
        private lateinit var eggs: Array<Egg>
    }

    private data class Egg(
        var hp: Int,
        val atk: Int,
    )

    fun solution() {
        N = readln().toInt()
        eggs = Array(N) { Egg(0, 0) }
        for (i in 0 until N) {
            val l = StringTokenizer(readln())
            val hp = l.nextToken().toInt()
            val atk = l.nextToken().toInt()
            eggs[i] = Egg(hp, atk)
        }

        dfs(0)
        print(max)
    }

    private fun dfs(cur: Int) {
        if (cur == N) {
            max = max(max, eggs.count { it.hp <= 0 })
            return
        }

        if (eggs[cur].hp <= 0) {
            dfs(cur + 1)
            return
        }

        for (next in 0 until N) {
            if (next == cur) continue
            if (eggs[next].hp > 0) {
                eggs[cur].hp -= eggs[next].atk
                eggs[next].hp -= eggs[cur].atk
                dfs(cur + 1)
                eggs[cur].hp += eggs[next].atk
                eggs[next].hp += eggs[cur].atk
            }
        }
        if (cur == N - 1) dfs(cur + 1)
    }
}

fun main() {
    Solution16987().solution()
}